cpp/src/arrow/parquet/writer.cc

```diff
@@ -143,6 +153,82 @@ Status FileWriter::Close() {
 
 FileWriter::~FileWriter() {}
 
+// Create a slice of a PrimitiveArray.
+//
+// This method is specially crafted for WriteFlatTable and assumes the following:
+//  * chunk_size is a multiple of 512
+Status TemporaryArraySlice(int64_t chunk, int64_t chunk_size, PrimitiveArray* array,
+    std::shared_ptr<PrimitiveArray>* out) {
+  // The last chunk may be smaller than the chunk_size
+  int64_t size = std::min(chunk_size, array->length() - chunk * chunk_size);
+  int64_t buffer_offset = chunk * chunk_size * array->type()->value_size();
+  int64_t value_size = size * array->type()->value_size();
+  auto chunk_buffer = std::make_shared<Buffer>(array->data(), buffer_offset, value_size);
+  std::shared_ptr<Buffer> null_bitmap;
+  int32_t null_count = 0;
+  if (array->null_count() > 0) {
+    int64_t null_offset = (chunk * chunk_size) / 8;
+    int64_t null_size = util::ceil_byte(size) / 8;
+    null_bitmap = std::make_shared<Buffer>(array->null_bitmap(), null_offset, null_size);
```

Comment:
oh, i see you need the null bitmap anyways down below.   I think think it might pay to convert the loop, so you can use the IsNull function directly from array.